File: results/0.75/PNG-Image/png-image-meta-llama-llama-3.3-70b-instruct-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include <stdlib.h>

#define UTF8(s) s, (sizeof(s) - 1)

typedef struct {
    uint32_t length;
    char type[4];
    uint8_t* data;
    uint32_t crc;
} chunk_t;

typedef struct {
    uint32_t width;
    uint32_t height;
    uint8_t bit_depth;
    uint8_t color_type;
    uint8_t compression_method;
    uint8_t filter_method;
    uint8_t interlace_method;
} ihdr_t;

typedef struct {
    uint8_t* palette;
    uint32_t length;
} plte_t;

typedef struct {
    uint8_t* data;
    uint32_t length;
} idat_t;

typedef struct {
    uint8_t intend;
} iend_t;

typedef struct {
    uint8_t* chromaticity;
    uint32_t length;
} chrm_t;

typedef struct {
    uint32_t gamma;
} gama_t;

typedef struct {
    uint8_t* significant_bits;
    uint32_t length;
} sbit_t;

typedef struct {
    uint8_t* srgb;
    uint32_t length;
} srgb_t;

typedef struct {
    uint8_t* iccp;
    uint32_t length;
} iccp_t;

typedef struct {
    uint8_t* text;
    uint32_t length;
    uint8_t* key;
    uint32_t key_length;
} text_t;

typedef struct {
    uint8_t* compressed_text;
    uint32_t length;
    uint8_t* key;
    uint32_t key_length;
} ztxt_t;

typedef struct {
    uint8_t* international_text;
    uint32_t length;
    uint8_t* key;
    uint32_t key_length;
    uint8_t* langtag;
    uint32_t langtag_length;
} itxt_t;

typedef struct {
    uint8_t* background;
    uint32_t length;
} bkgd_t;

typedef struct {
    uint8_t* histogram;
    uint32_t length;
} hist_t;

typedef struct {
    uint8_t* transparency;
    uint32_t length;
} trns_t;

typedef struct {
    uint32_t phys_units;
    uint32_t phys_ppu_x;
    uint32_t phys_ppu_y;
} phys_t;

typedef struct {
    int32_t scale_numerator;
    int32_t scale_denominator;
} scal_t;

typedef struct {
    uint8_t* valid;
    uint32_t length;
} vird_t;

uint32_t hammer_uint32_be(uint8_t* input, int* len) {
    uint32_t result = (input[0] << 24) | (input[1] << 16) | (input[2] << 8) | input[3];
    *len -= 4;
    return result;
}

uint8_t hammer_uint8(uint8_t* input, int* len) {
    uint8_t result = input[0];
    *len -= 1;
    return result;
}

char* hammer_string(uint8_t* input, int length, int* len) {
    char* result = malloc(length + 1);
    memcpy(result, input, length);
    result[length] = '\0';
    *len -= length;
    return result;
}

uint8_t* hammer_bytes(uint8_t* input, int length, int* len) {
    uint8_t* result = malloc(length);
    memcpy(result, input, length);
    *len -= length;
    return result;
}

int hammer_remaining(uint8_t* input, int* len) {
    return *len;
}

uint32_t hammer_uint32_be_input(uint8_t* input, int* len) {
    return hammer_uint32_be(input, len);
}

void* parse_chunk(uint8_t* input, void** output, int* len) {
    chunk_t* chunk = malloc(sizeof(chunk_t));
    *output = chunk;

    chunk->length = hammer_uint32_be(input, len);

    chunk->type[0] = input[0];
    chunk->type[1] = input[1];
    chunk->type[2] = input[2];
    chunk->type[3] = input[3];
    *len -= 4;

    chunk->data = hammer_bytes(input, chunk->length, len);

    chunk->crc = hammer_uint32_be(input, len);

    return NULL;
}

void* parse_ihdr(uint8_t* input, void** output, int* len) {
    ihdr_t* ihdr = malloc(sizeof(ihdr_t));
    *output = ihdr;

    ihdr->width = hammer_uint32_be(input, len);

    ihdr->height = hammer_uint32_be(input, len);

    ihdr->bit_depth = hammer_uint8(input, len);

    ihdr->color_type = hammer_uint8(input, len);

    ihdr->compression_method = hammer_uint8(input, len);

    ihdr->filter_method = hammer_uint8(input, len);

    ihdr->interlace_method = hammer_uint8(input, len);

    return NULL;
}

void* parse_plte(uint8_t* input, void** output, int* len) {
    plte_t* plte = malloc(sizeof(plte_t));
    *output = plte;

    plte->length = hammer_remaining(input, len);

    plte->palette = hammer_bytes(input, plte->length, len);

    return NULL;
}

void* parse_idat(uint8_t* input, void** output, int* len) {
    idat_t* idat = malloc(sizeof(idat_t));
    *output = idat;

    idat->length = hammer_remaining(input, len);

    idat->data = hammer_bytes(input, idat->length, len);

    return NULL;
}

void* parse_iend(uint8_t* input, void** output, int* len) {
    iend_t* iend = malloc(sizeof(iend_t));
    *output = iend;

    return NULL;
}

void* parse_chrm(uint8_t* input, void** output, int* len) {
    chrm_t* chrm = malloc(sizeof(chrm_t));
    *output = chrm;

    chrm->length = hammer_remaining(input, len);

    chrm->chromaticity = hammer_bytes(input, chrm->length, len);

    return NULL;
}

void* parse_gama(uint8_t* input, void** output, int* len) {
    gama_t* gama = malloc(sizeof(gama_t));
    *output = gama;

    gama->gamma = hammer_uint32_be(input, len);

    return NULL;
}

void* parse_sbit(uint8_t* input, void** output, int* len) {
    sbit_t* sbit = malloc(sizeof(sbit_t));
    *output = sbit;

    sbit->length = hammer_remaining(input, len);

    sbit->significant_bits = hammer_bytes(input, sbit->length, len);

    return NULL;
}

void* parse_srgb(uint8_t* input, void** output, int* len) {
    srgb_t* srgb = malloc(sizeof(srgb_t));
    *output = srgb;

    srgb->length = hammer_remaining(input, len);

    srgb->srgb = hammer_bytes(input, srgb->length, len);

    return NULL;
}

void* parse_iccp(uint8_t* input, void** output, int* len) {
    iccp_t* iccp = malloc(sizeof(iccp_t));
    *output = iccp;

    iccp->length = hammer_remaining(input, len);

    iccp->iccp = hammer_bytes(input, iccp->length, len);

    return NULL;
}

void* parse_text(uint8_t* input, void** output, int* len) {
    text_t* text = malloc(sizeof(text_t));
    *output = text;

    text->key_length = hammer_remaining(input, len);

    text->key = hammer_bytes(input, text->key_length, len);

    text->length = hammer_remaining(input, len);

    text->text = hammer_bytes(input, text->length, len);

    return NULL;
}

void* parse_ztxt(uint8_t* input, void** output, int* len) {
    ztxt_t* ztxt = malloc(sizeof(ztxt_t));
    *output = ztxt;

    ztxt->key_length = hammer_remaining(input, len);

    ztxt->key = hammer_bytes(input, ztxt->key_length, len);

    ztxt->length = hammer_remaining(input, len);

    ztxt->compressed_text = hammer_bytes(input, ztxt->length, len);

    return NULL;
}

void* parse_itxt(uint8_t* input, void** output, int* len) {
    itxt_t* itxt = malloc(sizeof(itxt_t));
    *output = itxt;

    itxt->key_length = hammer_remaining(input, len);

    itxt->key = hammer_bytes(input, itxt->key_length, len);

    itxt->langtag_length = hammer_remaining(input, len);

    itxt->langtag = hammer_bytes(input, itxt->langtag_length, len);

    itxt->length = hammer_remaining(input, len);

    itxt->international_text = hammer_bytes(input, itxt->length, len);

    return NULL;
}

void* parse_bkgd(uint8_t* input, void** output, int* len) {
    bkgd_t* bkgd = malloc(sizeof(bkgd_t));
    *output = bkgd;

    bkgd->length = hammer_remaining(input, len);

    bkgd->background = hammer_bytes(input, bkgd->length, len);

    return NULL;
}

void* parse_hist(uint8_t* input, void** output, int* len) {
    hist_t* hist = malloc(sizeof(hist_t));
    *output = hist;

    hist->length = hammer_remaining(input, len);

    hist->histogram = hammer_bytes(input, hist->length, len);

    return NULL;
}

void* parse_trns(uint8_t* input, void** output, int* len) {
    trns_t* trns = malloc(sizeof(trns_t));
    *output = trns;

    trns->length = hammer_remaining(input, len);

    trns->transparency = hammer_bytes(input, trns->length, len);

    return NULL;
}

void* parse_phys(uint8_t* input, void** output, int* len) {
    phys_t* phys = malloc(sizeof(phys_t));
    *output = phys;

    phys->phys_units = hammer_uint32_be(input, len);

    phys->phys_ppu_x = hammer_uint32_be(input, len);

    phys->phys_ppu_y = hammer_uint32_be(input, len);

    return NULL;
}

void* parse_scal(uint8_t* input, void** output, int* len) {
    scal_t* scal = malloc(sizeof(scal_t));
    *output = scal;

    scal->scale_numerator = (int32_t)hammer_uint32_be(input, len);

    scal->scale_denominator = (int32_t)hammer_uint32_be(input, len);

    return NULL;
}

void* parse_vird(uint8_t* input, void** output, int* len) {
    vird_t* vird = malloc(sizeof(vird_t));
    *output = vird;

    vird->length = hammer_remaining(input, len);

    vird->valid = hammer_bytes(input, vird->length, len);

    return NULL;
}

int main(int argc, char** argv) {
    if (argc != 2) {
        printf("Usage: %s <input_file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (file == NULL) {
        printf("Error opening file: %s\n", argv[1]);
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    uint8_t* file_data = malloc(file_size);
    fread(file_data, 1, file_size, file);
    fclose(file);

    uint8_t* input = file_data;
    int len = file_size;

    uint8_t* signature = hammer_bytes(input, 8, &len);
    if (memcmp(signature, "\x89\x50\x4E\x47\x0D\x0A\x1A\x0A", 8) != 0) {
        printf("Invalid PNG signature\n");
        return 1;
    }

    while (len > 0) {
        chunk_t* chunk;
        parse_chunk(input, (void**)&chunk, &len);

        if (strcmp(chunk->type, "IHDR") == 0) {
            ihdr_t* ihdr;
            parse_ihdr(chunk->data, (void**)&ihdr, &chunk->length);
            printf("IHDR chunk:\n");
            printf("  width: %u\n", ihdr->width);
            printf("  height: %u\n", ihdr->height);
            printf("  bit depth: %u\n", ihdr->bit_depth);
            printf("  color type: %u\n", ihdr->color_type);
            printf("  compression method: %u\n", ihdr->compression_method);
            printf("  filter method: %u\n", ihdr->filter_method);
            printf("  interlace method: %u\n", ihdr->interlace_method);
        } else if (strcmp(chunk->type, "PLTE") == 0) {
            plte_t* plte;
            parse_plte(chunk->data, (void**)&plte, &chunk->length);
            printf("PLTE chunk:\n");
            printf("  palette: %p\n", plte->palette);
        } else if (strcmp(chunk->type, "IDAT") == 0) {
            idat_t* idat;
            parse_idat(chunk->data, (void**)&idat, &chunk->length);
            printf("IDAT chunk:\n");
            printf("  data: %p\n", idat->data);
        } else if (strcmp(chunk->type, "IEND") == 0) {
            iend_t* iend;
            parse_iend(chunk->data, (void**)&iend, &chunk->length);
            printf("IEND chunk:\n");
        } else if (strcmp(chunk->type, "cHRM") == 0) {
            chrm_t* chrm;
            parse_chrm(chunk->data, (void**)&chrm, &chunk->length);
            printf("cHRM chunk:\n");
            printf("  chromaticity: %p\n", chrm->chromaticity);
        } else if (strcmp(chunk->type, "gAMA") == 0) {
            gama_t* gama;
            parse_gama(chunk->data, (void**)&gama, &chunk->length);
            printf("gAMA chunk:\n");
            printf("  gamma: %u\n", gama->gamma);
        } else if (strcmp(chunk->type, "sBIT") == 0) {
            sbit_t* sbit;
            parse_sbit(chunk->data, (void**)&sbit, &chunk->length);
            printf("sBIT chunk:\n");
            printf("  significant bits: %p\n", sbit->significant_bits);
        } else if (strcmp(chunk->type, "sRGB") == 0) {
            srgb_t* srgb;
            parse_srgb(chunk->data, (void**)&srgb, &chunk->length);
            printf("sRGB chunk:\n");
            printf("  srgb: %p\n", srgb->srgb);
        } else if (strcmp(chunk->type, "iCCP") == 0) {
            iccp_t* iccp;
            parse_iccp(chunk->data, (void**)&iccp, &chunk->length);
            printf("iCCP chunk:\n");
            printf("  iccp: %p\n", iccp->iccp);
        } else if (strcmp(chunk->type, "tEXt") == 0) {
            text_t* text;
            parse_text(chunk->data, (void**)&text, &chunk->length);
            printf("tEXt chunk:\n");
            printf("  key: %p\n", text->key);
            printf("  text: %p\n", text->text);
        } else if (strcmp(chunk->type, "zTXt") == 0) {
            ztxt_t* ztxt;
            parse_ztxt(chunk->data, (void**)&ztxt, &chunk->length);
            printf("zTXt chunk:\n");
            printf("  key: %p\n", ztxt->key);
            printf("  compressed text: %p\n", ztxt->compressed_text);
        } else if (strcmp(chunk->type, "iTXt") == 0) {
            itxt_t* itxt;
            parse_itxt(chunk->data, (void**)&itxt, &chunk->length);
            printf("iTXt chunk:\n");
            printf("  key: %p\n", itxt->key);
            printf("  language tag: %p\n", itxt->langtag);
            printf("  international text: %p\n", itxt->international_text);
        } else if (strcmp(chunk->type, "bKGD") == 0) {
            bkgd_t* bkgd;
            parse_bkgd(chunk->data, (void**)&bkgd, &chunk->length);
            printf("bKGD chunk:\n");
            printf("  background: %p\n", bkgd->background);
        } else if (strcmp(chunk->type, "hIST") == 0) {
            hist_t* hist;
            parse_hist(chunk->data, (void**)&hist, &chunk->length);
            printf("hIST chunk:\n");
            printf("  histogram: %p\n", hist->histogram);
        } else if (strcmp(chunk->type, "tRNS") == 0) {
            trns_t* trns;
            parse_trns(chunk->data, (void**)&trns, &chunk->length);
            printf("tRNS chunk:\n");
            printf("  transparency: %p\n", trns->transparency);
        } else if (strcmp(chunk->type, "pHYs") == 0) {
            phys_t* phys;
            parse_phys(chunk->data, (void**)&phys, &chunk->length);
            printf("pHYs chunk:\n");
            printf("  phys units: %u\n", phys->phys_units);
            printf("  phys ppu x: %u\n", phys->phys_ppu_x);
            printf("  phys ppu y: %u\n", phys->phys_ppu_y);
        } else if (strcmp(chunk->type, "sCAL") == 0) {
            scal_t* scal;
            parse_scal(chunk->data, (void**)&scal, &chunk

File: results/0.75/PNG-Image/png-image-meta-llama-llama-3.3-70b-instruct-turbo.rs
use nom::{
    bytes::complete::{tag, take},
    combinator::{map, map_res},
    error::{Error, ErrorKind},
    multi::{many0, many1},
    number::complete::{be_u16, be_u32, be_u8},
    sequence::{delimited, tuple},
    IResult,
};
use std::{
    env,
    fs::File,
    io::{BufReader, Read},
};

#[derive(Debug, PartialEq)]
enum ColorType {
    Grayscale,
    RGB,
    Palette,
    GreyscaleAlpha,
    RGBA,
}

#[derive(Debug, PartialEq)]
enum InterlaceMethod {
    NoInterlace,
    Adam7,
}

#[derive(Debug, PartialEq)]
enum CompressionMethod {
    DeflateInflate,
}

#[derive(Debug, PartialEq)]
enum FilterMethod {
    AdaptiveFiltering,
}

#[derive(Debug, PartialEq)]
struct IHDR {
    width: u32,
    height: u32,
    bit_depth: u8,
    color_type: ColorType,
    compression_method: CompressionMethod,
    filter_method: FilterMethod,
    interlace_method: InterlaceMethod,
}

#[derive(Debug, PartialEq)]
struct PLTE {
    palette_entries: Vec<(u8, u8, u8)>,
}

#[derive(Debug, PartialEq)]
struct IDAT {
    compressed_data: Vec<u8>,
}

#[derive(Debug, PartialEq)]
struct IEND {}

#[derive(Debug, PartialEq)]
struct tRNS {
    transparency_values: Vec<u8>,
}

#[derive(Debug, PartialEq)]
struct cHRM {
    white_point_x: u32,
    white_point_y: u32,
    red_x: u32,
    red_y: u32,
    green_x: u32,
    green_y: u32,
    blue_x: u32,
    blue_y: u32,
}

#[derive(Debug, PartialEq)]
struct gAMA {
    gamma_value: u32,
}

#[derive(Debug, PartialEq)]
struct iCCP {
    profile_name: String,
    compression_method: CompressionMethod,
    compressed_profile: Vec<u8>,
}

#[derive(Debug, PartialEq)]
struct sBIT {
    significant_bits: (u8, u8, u8),
}

#[derive(Debug, PartialEq)]
struct sRGB {
    rendering_intent: u8,
}

#[derive(Debug, PartialEq)]
struct tXYZ {
    transformation_matrix: [(i32, i32, i32); 3],
}

#[derive(Debug, PartialEq)]
struct bKGD {
    background_color: (u8, u8, u8),
}

#[derive(Debug, PartialEq)]
struct hIST {
    histogram_entries: Vec<u16>,
}

#[derive(Debug, PartialEq)]
struct pHYs {
    pixels_per_unit_x: u32,
    pixels_per_unit_y: u32,
    unit_specifier: u8,
}

#[derive(Debug, PartialEq)]
struct sCAL {
    unit_specifier: u8,
    width: u32,
    height: u32,
}

#[derive(Debug, PartialEq)]
struct tEXt {
    keyword: String,
    text_data: String,
}

#[derive(Debug, PartialEq)]
struct zTXt {
    keyword: String,
    compression_method: CompressionMethod,
    compressed_text_data: Vec<u8>,
}

#[derive(Debug, PartialEq)]
struct iTXt {
    keyword: String,
    compression_flag: bool,
    compression_method: CompressionMethod,
    language_tag: String,
    translated_keyword: String,
    text_data: String,
}

#[derive(Debug, PartialEq)]
struct fRAc {
    numerator: u32,
    denominator: u32,
}

fn magic_number(input: &[u8]) -> IResult<&[u8], ()> {
    tag(&[0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a])(input).map(|(input, _)| (input, ()))
}

fn ihdr(input: &[u8]) -> IResult<&[u8], IHDR> {
    let (input, _) = tag("IHDR")(input)?;
    let (input, width) = be_u32(input)?;
    let (input, height) = be_u32(input)?;
    let (input, bit_depth) = be_u8(input)?;
    let (input, color_type) = map(be_u8, |color_type| match color_type {
        0 => ColorType::Grayscale,
        2 => ColorType::RGB,
        3 => ColorType::Palette,
        4 => ColorType::GreyscaleAlpha,
        6 => ColorType::RGBA,
        _ => panic!("Invalid color type"),
    })(input)?;
    let (input, _) = be_u8(input)?; // compression method
    let (input, _) = be_u8(input)?; // filter method
    let (input, interlace_method) = map(be_u8, |interlace_method| match interlace_method {
        0 => InterlaceMethod::NoInterlace,
        1 => InterlaceMethod::Adam7,
        _ => panic!("Invalid interlace method"),
    })(input)?;
    Ok((input, IHDR {
        width,
        height,
        bit_depth,
        color_type,
        compression_method: CompressionMethod::DeflateInflate,
        filter_method: FilterMethod::AdaptiveFiltering,
        interlace_method,
    }))
}

fn plte(input: &[u8]) -> IResult<&[u8], PLTE> {
    let (input, _) = tag("PLTE")(input)?;
    let (input, palette_entries) = many1(tuple((be_u8, be_u8, be_u8)))(input)?;
    Ok((input, PLTE { palette_entries }))
}

fn idat(input: &[u8]) -> IResult<&[u8], IDAT> {
    let (input, _) = tag("IDAT")(input)?;
    let (input, compressed_data) = take(4)(input)?;
    Ok((input, IDAT { compressed_data: compressed_data.to_vec() }))
}

fn iend(input: &[u8]) -> IResult<&[u8], IEND> {
    let (input, _) = tag("IEND")(input)?;
    Ok((input, IEND {}))
}

fn trns(input: &[u8]) -> IResult<&[u8], tRNS> {
    let (input, _) = tag("tRNS")(input)?;
    let (input, transparency_values) = many1(be_u8)(input)?;
    Ok((input, tRNS { transparency_values }))
}

fn chrm(input: &[u8]) -> IResult<&[u8], cHRM> {
    let (input, _) = tag("cHRM")(input)?;
    let (input, white_point_x) = be_u32(input)?;
    let (input, white_point_y) = be_u32(input)?;
    let (input, red_x) = be_u32(input)?;
    let (input, red_y) = be_u32(input)?;
    let (input, green_x) = be_u32(input)?;
    let (input, green_y) = be_u32(input)?;
    let (input, blue_x) = be_u32(input)?;
    let (input, blue_y) = be_u32(input)?;
    Ok((input, cHRM {
        white_point_x,
        white_point_y,
        red_x,
        red_y,
        green_x,
        green_y,
        blue_x,
        blue_y,
    }))
}

fn gama(input: &[u8]) -> IResult<&[u8], gAMA> {
    let (input, _) = tag("gAMA")(input)?;
    let (input, gamma_value) = be_u32(input)?;
    Ok((input, gAMA { gamma_value }))
}

fn iccp(input: &[u8]) -> IResult<&[u8], iCCP> {
    let (input, _) = tag("iCCP")(input)?;
    let (input, profile_name) = map_res(take(79), |profile_name| String::from_utf8_lossy(profile_name).into_owned())(input)?;
    let (input, _) = be_u8(input)?; // compression method
    let (input, compressed_profile) = take(4)(input)?;
    Ok((input, iCCP {
        profile_name,
        compression_method: CompressionMethod::DeflateInflate,
        compressed_profile: compressed_profile.to_vec(),
    }))
}

fn sbit(input: &[u8]) -> IResult<&[u8], sBIT> {
    let (input, _) = tag("sBIT")(input)?;
    let (input, significant_bits) = tuple((be_u8, be_u8, be_u8))(input)?;
    Ok((input, sBIT { significant_bits }))
}

fn srgb(input: &[u8]) -> IResult<&[u8], sRGB> {
    let (input, _) = tag("sRGB")(input)?;
    let (input, rendering_intent) = be_u8(input)?;
    Ok((input, sRGB { rendering_intent }))
}

fn txyz(input: &[u8]) -> IResult<&[u8], tXYZ> {
    let (input, _) = tag("tXYZ")(input)?;
    let (input, transformation_matrix) = many1(tuple((be_u32, be_u32, be_u32)))(input)?;
    Ok((input, tXYZ {
        transformation_matrix: transformation_matrix
            .into_iter()
            .map(|(a, b, c)| (a as i32, b as i32, c as i32))
            .collect::<Vec<_>>()
            .try_into()
            .unwrap(),
    }))
}

fn bkgd(input: &[u8]) -> IResult<&[u8], bKGD> {
    let (input, _) = tag("bKGD")(input)?;
    let (input, background_color) = tuple((be_u8, be_u8, be_u8))(input)?;
    Ok((input, bKGD { background_color }))
}

fn hist(input: &[u8]) -> IResult<&[u8], hIST> {
    let (input, _) = tag("hIST")(input)?;
    let (input, histogram_entries) = many1(be_u16)(input)?;
    Ok((input, hIST { histogram_entries }))
}

fn phys(input: &[u8]) -> IResult<&[u8], pHYs> {
    let (input, _) = tag("pHYs")(input)?;
    let (input, pixels_per_unit_x) = be_u32(input)?;
    let (input, pixels_per_unit_y) = be_u32(input)?;
    let (input, unit_specifier) = be_u8(input)?;
    Ok((input, pHYs {
        pixels_per_unit_x,
        pixels_per_unit_y,
        unit_specifier,
    }))
}

fn scal(input: &[u8]) -> IResult<&[u8], sCAL> {
    let (input, _) = tag("sCAL")(input)?;
    let (input, unit_specifier) = be_u8(input)?;
    let (input, width) = be_u32(input)?;
    let (input, height) = be_u32(input)?;
    Ok((input, sCAL { unit_specifier, width, height }))
}

fn tuxt(input: &[u8]) -> IResult<&[u8], tEXt> {
    let (input, _) = tag("tEXt")(input)?;
    let (input, keyword) = map_res(take(79), |keyword| String::from_utf8_lossy(keyword).into_owned())(input)?;
    let (input, text_data) = map_res(take(79), |text_data| String::from_utf8_lossy(text_data).into_owned())(input)?;
    Ok((input, tEXt { keyword, text_data }))
}

fn ztzt(input: &[u8]) -> IResult<&[u8], zTXt> {
    let (input, _) = tag("zTXt")(input)?;
    let (input, keyword) = map_res(take(79), |keyword| String::from_utf8_lossy(keyword).into_owned())(input)?;
    let (input, _) = be_u8(input)?; // compression method
    let (input, compressed_text_data) = take(4)(input)?;
    Ok((input, zTXt {
        keyword,
        compression_method: CompressionMethod::DeflateInflate,
        compressed_text_data: compressed_text_data.to_vec(),
    }))
}

fn itxt(input: &[u8]) -> IResult<&[u8], iTXt> {
    let (input, _) = tag("iTXt")(input)?;
    let (input, keyword) = map_res(take(79), |keyword| String::from_utf8_lossy(keyword).into_owned())(input)?;
    let (input, compression_flag) = map(be_u8, |compression_flag| compression_flag != 0)(input)?;
    let (input, _) = be_u8(input)?; // compression method
    let (input, language_tag) = map_res(take(79), |language_tag| String::from_utf8_lossy(language_tag).into_owned())(input)?;
    let (input, translated_keyword) = map_res(take(79), |translated_keyword| String::from_utf8_lossy(translated_keyword).into_owned())(input)?;
    let (input, text_data) = map_res(take(79), |text_data| String::from_utf8_lossy(text_data).into_owned())(input)?;
    Ok((input, iTXt {
        keyword,
        compression_flag,
        compression_method: CompressionMethod::DeflateInflate,
        language_tag,
        translated_keyword,
        text_data,
    }))
}

fn frac(input: &[u8]) -> IResult<&[u8], fRAc> {
    let (input, _) = tag("fRAc")(input)?;
    let (input, numerator) = be_u32(input)?;
    let (input, denominator) = be_u32(input)?;
    Ok((input, fRAc { numerator, denominator }))
}

fn png(input: &[u8]) -> IResult<&[u8], Vec<Box<dyn std::any::Any>>> {
    let (input, _) = magic_number(input)?;
    let mut chunks = Vec::new();
    loop {
        let (input, chunk_type) = take(4u8)(input)?;
        match chunk_type {
            b"IHDR" => {
                let (input, ihdr) = ihdr(input)?;
                chunks.push(Box::new(ihdr));
            }
            b"PLTE" => {
                let (input, plte) = plte(input)?;
                chunks.push(Box::new(plte));
            }
            b"IDAT" => {
                let (input, idat) = idat(input)?;
                chunks.push(Box::new(idat));
            }
            b"IEND" => {
                let (input, _) = iend(input)?;
                break;
            }
            b"tRNS" => {
                let (input, trns) = trns(input)?;
                chunks.push(Box::new(trns));
            }
            b"cHRM" => {
                let (input, chrm) = chrm(input)?;
                chunks.push(Box::new(chrm));
            }
            b"gAMA" => {
                let (input, gama) = gama(input)?;
                chunks.push(Box::new(gama));
            }
            b"iCCP" => {
                let (input, iccp) = iccp(input)?;
                chunks.push(Box::new(iccp));
            }
            b"sBIT" => {
                let (input, sbit) = sbit(input)?;
                chunks.push(Box::new(sbit));
            }
            b"sRGB" => {
                let (input, srgb) = srgb(input)?;
                chunks.push(Box::new(srgb));
            }
            b"tXYZ" => {
                let (input, txyz) = txyz(input)?;
                chunks.push(Box::new(txyz));
            }
            b"bKGD" => {
                let (input, bkgd) = bkgd(input)?;
                chunks.push(Box::new(bkgd));
            }
            b"hIST" => {
                let (input, hist) = hist(input)?;
                chunks.push(Box::new(hist));
            }
            b"pHYs" => {
                let (input, phys) = phys(input)?;
                chunks.push(Box::new(phys));
            }
            b"sCAL" => {
                let (input, scal) = scal(input)?;
                chunks.push(Box::new(scal));
            }
            b"tEXt" => {
                let (input, tuxt) = tuxt(input)?;
                chunks.push(Box::new(tuxt));
            }
            b"zTXt" => {
                let (input, ztzt) = ztzt(input)?;
                chunks.push(Box::new(ztzt));
            }
            b"iTXt" => {
                let (input, itxt) = itxt(input)?;
                chunks.push(Box::new(itxt));
            }
            b"fRAc" => {
                let (input, frac) = frac(input)?;
                chunks.push(Box::new(frac));
            }
            _ => {
                let (input, _) = take(4)(input)?;
                continue;
            }
        }
    }
    Ok((input, chunks))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Usage: {} <input_file>", args[0]);
    }
    let file = File::open(&args[1]).unwrap();
    let mut reader = BufReader::new(file);
    let mut input = Vec::new();
    reader.read_to_end(&mut input).unwrap();
    let (_rest, chunks) = png(&input).unwrap();
    for chunk in chunks {
        match chunk.downcast_ref::<IHDR>() {
            Some(ihdr) => println!("IHDR: {:?}", ihdr),
            None => (),
        }
        match chunk.downcast_ref::<PLTE>() {
            Some(plte) => println!("PLTE: {:?}", plte),
            None => (),
        }
        match chunk.downcast_ref::<IDAT>() {
            Some(idat) => println!("IDAT: {:?}", idat),
            None => (),
        }
        match chunk.downcast_ref::<IEND>() {
            Some(_) => println!("IEND"),
            None => (),
        }
        match chunk.downcast_ref::<tRNS>() {
            Some(trns) => println!("tRNS: {:?}", trns),
            None => (),
        }
        match chunk.downcast_ref::<cHRM>() {
            Some(chrm) => println!("cHRM: {:?}", chrm),
            None => (),
        }
        match chunk.downcast_ref::<gAMA>() {
            Some(gama) => println!("gAMA: {:?}", gama),
            None => (),
        }
        match chunk.downcast_ref::<iCCP>() {
            Some(iccp) => println!("iCCP: {:?}", iccp),
            None => (),
        }
        match chunk.downcast_ref::<sBIT>() {
            Some(sbit) => println!("sBIT: {:?}", sbit),
            None => (),
